### DISCUSSION TITLE
How to properly use SwiftUI's .task modifier?

### DISCUSSION THREAD

**Discussion by kakhaberikiknadze** [pointfreeco/swift-composable-architecture#1829] [Q&A] [answered]
So, I've just noticed I'm getting this runtime warning

```
The "Store" class is not thread-safe, and so all interactions with an instance of "Store" (including all of its scopes and derived view stores) must be done on the main thread.
"ViewStore.send" was called on a non-main thread with: AppReducer.Action.foo …
```

Well, if I wrap `viewStore.send(.refresh)` inside `DispatchQueue.main.async` then it will be called on main thread and the warning goes away. 

Just wanted to make sure this is a correct implementation for running some background tasks when you navigate to the view.

``` Swift
.task {
    DispatchQueue.main.async {
        viewStore.send(.foo)
    }
}
```

**Comment by kakhaberikiknadze** [comment, marked as answer]
I've noticed I'm getting the same warning when I send the action from `refreshable` as well.

```
.refreshable {
    await viewStore.send(.refresh)
}
```

**Reply by mbrandonw** [reply]
Hi @kakhaberikiknadze, can you share a demo that reproduces this problem? We have case studies that use both [`.refreshable`](https://github.com/pointfreeco/swift-composable-architecture/blob/9e83d195fbc6c3a58a35fb11de01c6a2947b4935/Examples/CaseStudies/SwiftUICaseStudies/02-Effects-Refreshable.swift#L107) and [`.task`](https://github.com/pointfreeco/swift-composable-architecture/blob/9e83d195fbc6c3a58a35fb11de01c6a2947b4935/Examples/CaseStudies/SwiftUICaseStudies/02-Effects-LongLiving.swift#L91), and neither have this problem.

You can also put a breakpoint in each of those closures to confirm that it really is being called on a non-main thread. If it is not, then in means your problem is somewhere else, most likely an effect delivering an action on a non-main thread.

Or, if those closures really are being called on a non-main thread, you can always do:

```swift
.task { @MainActor in 
  ...
}
```

**Reply by kakhaberikiknadze** [reply]
Yup, I've debugged and they're indeed being called on non-main thread and also noticed that they're always called on main thread on the other view I've got. I'll try to setup a reproducible example and share. Meanwhile, thanks for the `@MainActor` suggestion.

**Reply by kakhaberikiknadze** [reply]
Oh, I've just found out exactly when it happens. It happens when you split the rendering code into dedicated methods and inject `viewStore` in them. Let's say, you add `.task` modifiers in both places. If you debug, you'll see the one inside `WithViewStore` is called on main thread whereas the other's always called on background thread.

``` Swift
struct MyView: View {
    let store: StoreOf<AppReducer>

    var body: some View {
        WithViewStore(store, observe: { $0 }) { viewStore in
            renderContent(viewStore)
                .task {
                    viewStore.send(.refresh)
                }
        }
    }
    
    func renderContent(_ viewStore: ViewStoreOf<AppReducer>) -> some View {
        Text("Foo")
            .task {
                viewStore.send(.refresh)
            }
    }
}
```

**Reply by kakhaberikiknadze** [reply]
The solution is to mark `func renderContent` as `@MainActor func renderContent`

**Reply by kakhaberikiknadze** [reply]
Or perhaps the struct itself to `@MainActor`

**Reply by mbrandonw** [reply]
Ah yes, you've run into the classic and subtle edge case of "main actor inference" for views. 😕 Read this [article](https://oleb.net/2022/swiftui-task-mainactor/) for more info.

**Reply by kakhaberikiknadze** [reply]
Thanks 🙌🏻 And actually, double-thanks for such quick responses 🚀

**Reply by tgrapperon** [reply]
Yes, there is a very tricky `MainActor` inheritance behavior in `View`. There is a nice explanation [here](https://oleb.net/2022/swiftui-task-mainactor/). When you wrote your helper functions, you lost the implicit `MainActor`, but you would had it if your view hosted a `@StateObject` property for example.
Edit: Ah, sorry, I didn't see Brandon's response, I was on mobile.
